File: flyweight/MainPlayer.java
package flyweight;

public class MainPlayer implements Player {

    private final String displayName;
    private final int attackPower = 5;
    private final int healthBar = 100;
    private Weapon weapon;

    public MainPlayer(String displayName) {
        System.out.println("Creating Main Player...");
        this.displayName = displayName;
    }

    public String getDisplayName() {
        return displayName;
    }

    public int getAttackPower() {
        return attackPower;
    }


    public int getHealthBar() {
        return healthBar;
    }

    public Weapon getWeapon() {
        return weapon;
    }

    public void setWeapon(Weapon weapon) {
        this.weapon = weapon;
    }

    @Override
    public void attack() {
        System.out.println("Player: " + displayName + " is attacking with weapon: "
                + weapon.getName() + " with base attack damage: " + attackPower);
    }

    @Override
    public void assignWeapon(Weapon weapon) {
        this.weapon = weapon;
    }
}

File: C#/Flyweight/MainPlayer.cs
namespace Flyweight;

public sealed class MainPlayer : IPlayer
{
    public string DisplayName { get; }
    public int AttackPower { get; } = 5;
    public int HealthBar { get; } = 100;
    public Weapon Weapon { get; set; }

    public MainPlayer(string displayName)
    {
        Console.WriteLine("Creating main player...");
        DisplayName = displayName;
    }

    public void AssignWeapon(Weapon weapon)
        => Weapon = weapon;

    public void Attack()
        => Console.WriteLine($"Player: {DisplayName} is attacking with weapon: {Weapon.Name} with base attack damage: {AttackPower}");
}
